src/libraries/Microsoft.PowerFx.Interpreter/Functions/Mutation/RemoveFunction.cs

```diff
@@ -0,0 +1,218 @@
+                    if (argCount >= 3 && i == argCount - 1 && DType.String.Accepts(argType))
+                    {
+                        var strNode = (StrLitNode)args[i];
+
+                        if (strNode.Value.ToUpper() != "ALL")
+                        {
+                            fValid = false;
+                            errors.EnsureError(args[i], ErrRemoveAllArg, args[i]);
+                        }
+
+                        continue;
+                    }
+
+                    fValid = false;
+                    errors.EnsureError(args[i], ErrNeedRecord, args[i]);
+                    continue;
+                }
+
+                fValid &= DropAttachmentsIfExists(ref argType, errors, args[i]);
+
+                var collectionAcceptsRecord = collectionType.Accepts(argType.ToTable());
+                var recordAcceptsCollection = argType.ToTable().Accepts(collectionType);
+
+                // The item schema should be compatible with the collection schema.
+                if (!collectionAcceptsRecord && !recordAcceptsCollection)
+                {
+                    fValid = false;
+                    if (!SetErrorForMismatchedColumns(collectionType, argType, args[i], errors))
+                    {
+                        errors.EnsureError(DocumentErrorSeverity.Severe, args[i], ErrTableDoesNotAcceptThisType);
+                    }
+                }
+
+                // Only warn about no-op record inputs if there are no data sources that would use reference identity for comparison.
+                else if (!collectionType.AssociatedDataSources.Any() && !recordAcceptsCollection)
+                {
+                    errors.EnsureError(DocumentErrorSeverity.Warning, args[i], ErrTableDoesNotAcceptThisType);
+                }
+            }
+
+            // Remove returns the new collection, so the return schema is the same as the collection schema.
+            returnType = collectionType;
+
+            return fValid;
+        }
+
+        public async Task<FormulaValue> InvokeAsync(FormulaValue[] args, CancellationToken cancel)
+        {
+            var validArgs = CheckArgs(args, out FormulaValue faultyArg);
+
+            if (!validArgs)
+            {
+                return faultyArg;
+            }
+
+            if (args[0] is BlankValue)
+            {
+                return args[0];
+            }
+
+            var argCount = args.Count();
+            var lastArg = args.Last() as FormulaValue;
+            var all = false;
+            var toExclude = 1;
+
+            if (argCount >= 3 && DType.String.Accepts(lastArg.Type._type))
+            {
+                var lastArgValue = (string)lastArg.ToObject();
+
+                if (lastArgValue.ToUpper() == "ALL")
+                {
+                    all = true;
+                    toExclude = 2;
+                }
+            }
+
+            var datasource = (TableValue)args[0];
+
+            // !JYL! Better approach?
```

Comment:
> // !JYL! Better approach? [](http://example.com/codeflow?start=11&length=26)

I think this is actually ok. 
It's still a O[n], traversing the enumeration once. 

The alternative is probably a call to Array.Copy
[https://docs.microsoft.com/en-us/dotnet/api/system.array.copy?view=net-6.0](https://docs.microsoft.com/en-us/dotnet/api/system.array.copy?view=net-6.0)